ansible_base/authentication/utils/claims.py

```diff
@@ -352,58 +401,279 @@ def create_organizations_and_teams(results) -> None:
-            return
-        for role in user.has_roles.filter(role_definition=rd):
-            rd.remove_permission(user, role.content_object)
-
     @classmethod
     def reconcile_user_claims(cls, user: AbstractUser, authenticator_user: AuthenticatorUser) -> None:
         logger.info("Reconciling user claims")
 
         claims = getattr(user, 'claims', authenticator_user.claims)
 
-        remove_users = authenticator_user.provider.remove_users
-        if remove_users:
-            cls._remove_role_from_user(user, RoleDefinition.objects.managed.org_member)
-            cls._remove_role_from_user(user, RoleDefinition.objects.managed.team_member)
+        if 'ansible_base.rbac' in settings.INSTALLED_APPS:
+            cls(claims, user, authenticator_user).manage_permissions()
+        else:
+            logger.info(_("Skipping user claims with RBAC roles, because RBAC app is not installed"))
+
+    def __init__(self, claims: dict, user: AbstractUser, authenticator_user: AuthenticatorUser):
+        """
+        :param claims: initialized by method create_claims()
+        """
+        self.authenticator_user = authenticator_user
+        self.claims = claims
+        self.permissions_cache = RoleUserAssignmentsCache()
+        self.rebuild_user_permissions = self.authenticator_user.provider.remove_users
+        self.user = user
+
+    def manage_permissions(self) -> None:
+        """Processes the user claims (key `rbac_roles`)
+        and adds/removes RBAC permissions (a.k.a. role_user_assignments)
+        """
+        self.permissions_cache.cache_existing(self.user.role_assignments.all())
+
+        # System roles
+        self._compute_system_permissions()
+
+        # Organization roles
+        for org, org_teams_dict in self._compute_organization_permissions():
+            # Team roles
+            self._compute_team_permissions(org, org_teams_dict)
+
+        self.apply_permissions()
+
+        # Superuser "role"
+        self.apply_superuser_permission()
+
+    def _compute_system_permissions(self) -> None:
+        for role_name, has_permission in self.claims['rbac_roles'].get('system', {}).get('roles', {}).items():
+            if has_permission:
+                self.permissions_cache.add(role_name, organization=None, team=None)
+            else:
+                self.permissions_cache.remove(role_name, organization=None, team=None)
+
+    def _compute_organization_permissions(self) -> Iterator[Tuple[AbstractOrganization, dict]]:
+        orgs_by_name = self._get_orgs_by_name(self.claims['rbac_roles'].get('organizations', {}).keys())
+
+        for org_name, org_details in self.claims['rbac_roles'].get('organizations', {}).items():
+            if (org := orgs_by_name.get(org_name)) is None:
+                logger.info(_("Skipping organization '{organization}', because the organization does not exist").format(organization=org_name))
+                continue
```

Comment:
We need to build the objects if they don't exist.

Reply:
the same, we may increase log severity instead, because this line shouldn't happen